docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,491 @@
+# EP: Definition of the Fleet and Device API spec
+<!-- this format is inspired by the K8S KEP format https://raw.githubusercontent.com/kubernetes/enhancements/master/keps/NNNN-kep-template/README.md -->
+## Summary
+
+The Device API, and Fleet API, and the dynamics of managing individual
+devices as well as fleets of devices is a fundamental piece of flightctl.
+
+The Fleet API must serve as a way to manage groups of devices while still allowing
+for the management of individual devices in a way that we can integrate with other
+management systems.
+
+## Motivation
+
+We need agreement on the API design direction and the semantics of the API to be able to
+implement and grow the flightctl service.
+
+### Goals
+
+* Define the format and behavior of the Device and Fleet APIs
+
+* Allow the management of individual fleetless devices on a smaller scale,
+  as an integration mechanism for services like AAP and ACM.
+
+* Allow the management of fleets of devices on a larger scale, via gitops or
+  API.
+
+* Change API fields to provide a more consistent experience.
+
+* Enable temporary/manual drift of fleet-managed devices for debug purposes.
+
+### Non-Goals
+
+
+
+## Proposal
+
+### Device API
+
+The device API is divided into multiple methods, some user facing, and other device/facing,
+this design facilitates cleaner RBAC control of the endpoint access. The diagram is only
+intended to illustrate the flow of information between the user/system and the device.
+
+```mermaid
+block-beta
+columns 1
+  user(["User"])
+
+  space
+
+
+  block
+    frontEndpoint["/device/{name}"]
+  end
+
+  user --> frontEndpoint
+  frontEndpoint --> user
+
+  space
+  block
+    metadata
+    devSpec["spec"]
+    devStatus["status"]
+  end
+
+  space
+
+  frontEndpoint --> metadata
+  frontEndpoint --> devSpec
+
+  block
+     renderedEndpoint["/device/{name}/rendered"]
+     statusEndpoint["/device/{name}/status"]
+  end
+
+  space
+
+  statusEndpoint --> devStatus
+  statusEndpoint --> devStatus
+  devSpec --> renderedEndpoint
+  renderedEndpoint --> Device
+  Device --> statusEndpoint
+
+
+  Device(["Device"])
+
+  style frontEndpoint fill:#969,stroke:#333,stroke-width:4px
+  style renderedEndpoint fill:#969,stroke:#333,stroke-width:4px
+  style statusEndpoint fill:#969,stroke:#333,stroke-width:4px
+  ```
+
+#### `/device/{name}` User facing method
+
+This endpoint can be accessed by the user or external systems to
+manage individual devices, also internally from the fleet specifications
+for devices.
+
+The proposed changes include:
+
+* Reincorporating device.spec as a visible part through this frontend API.
+
+* The `device.spec.os.image` contains a reference to the image to be used for the device.
+
+* The `device.spec.templateVersion` field is moved to `device.metadata.resourceVersion` opaque string identifies
```

Comment:
In K8s, `resourceVersion` refers to the version of the resource itself (here: the Device), while we want to record the version of the `fleet.spec.template` used to generate the `device.spec`. So that may be confusing.
Also, isn't this a record that the fleet controller keeps about the device? In this case, I believe an annotation would be a better fit, i.e. make this `device.metadata.annotations["fleet-controller/templateVersion"]=...`?

Reply:
It could make sense also in terms of fully decoupling any Device-> Fleet dependency, while Fleet-->Device Deps are ok, but:

There will be some coupling anyway, for example when we ask for a rendered device, and we provide a past "resourceVersion", to only return on a diff...  https://github.com/flightctl/flightctl/pull/247/files#diff-d194c209c6868d086baff7646aceb9c013396c93149640138a62b12ed02a834cR331

The device API needs to know about this to be resolved.

Asking @avishayt for a quick peek

Reply:
I agree with @fzdarsky.  Is it better than keeping it in the spec?

Reply:
I feel we're conflating two problems / mechanisms here:
1) `.metadata.resourceVersion` as a mechanism to ensure consistency (detect read-modify-write issues) and to get data selectively (see [K8s API](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions)). This is the opaque string the agent sends to the service in a param that should be called `resourceVersion` in GET (or maybe even HEAD) requests.
2) The mechanism for the fleet controller to store in a device the version of the _device template_ it used when rendering the device spec the last time. That part shouldn't have a definition in the Device API itself, so an annotation set by the fleet controller is a more natural fit.

Reply:
I think it's true, we are conflating it. But because the nature of resourceVersion being opaque, we can use it to track the template that was used to render this device, that would just be an implementation detail.


I will keep both, and make the explanation of what "resourceVersion" is more opaque, so we can use the exact same field and mechanism (during GET) as kubernetes does, while still track the original version of the fleet template in the annotation. Depending on the implementation those could match or not, but would leave room for divergence, and also, the same mechanism would be available to use when fleet controller isn't in place.

I'll update and let's see if it all sits in place or if it doesn't.

Reply:
Yes, resourceVersion is opaque, but for opportunistic concurrency control to work, it must be updated on every write to the device resource (incl. label updates), irrespective of whether the fleet's device template changed. So we cannot use it to track the template.